client/allocator/src/freeing_bump.rs

```diff
@@ -538,35 +572,42 @@ impl FreeingBumpHeapAllocator {
 /// accessible up to the reported size.
 ///
 /// The linear memory can grow in size with the wasm page granularity (64KiB), but it cannot shrink.
-pub trait Memory {
+trait MemoryExt {
 	/// Read a u64 from the heap in LE form. Returns an error if any of the bytes read are out of
 	/// bounds.
 	fn read_le_u64(&self, ptr: u32) -> Result<u64, Error>;
 	/// Write a u64 to the heap in LE form. Returns an error if any of the bytes written are out of
 	/// bounds.
 	fn write_le_u64(&mut self, ptr: u32, val: u64) -> Result<(), Error>;
 	/// Returns the full size of the memory in bytes.
-	fn size(&self) -> u32;
+	fn size(&self) -> u64;
 }
 
-impl Memory for [u8] {
+impl<T: Memory> MemoryExt for T {
 	fn read_le_u64(&self, ptr: u32) -> Result<u64, Error> {
-		let range =
-			heap_range(ptr, 8, self.len()).ok_or_else(|| error("read out of heap bounds"))?;
-		let bytes = self[range]
-			.try_into()
-			.expect("[u8] slice of length 8 must be convertible to [u8; 8]");
-		Ok(u64::from_le_bytes(bytes))
+		self.with_access(|memory| {
+			let range =
+				heap_range(ptr, 8, memory.len()).ok_or_else(|| error("read out of heap bounds"))?;
+			let bytes = memory[range]
+				.try_into()
+				.expect("[u8] slice of length 8 must be convertible to [u8; 8]");
+			Ok(u64::from_le_bytes(bytes))
+		})
 	}
+
 	fn write_le_u64(&mut self, ptr: u32, val: u64) -> Result<(), Error> {
-		let range =
-			heap_range(ptr, 8, self.len()).ok_or_else(|| error("write out of heap bounds"))?;
-		let bytes = val.to_le_bytes();
-		self[range].copy_from_slice(&bytes[..]);
-		Ok(())
+		self.with_access_mut(|memory| {
+			let range = heap_range(ptr, 8, memory.len())
+				.ok_or_else(|| error("write out of heap bounds"))?;
+			let bytes = val.to_le_bytes();
+			memory[range].copy_from_slice(&bytes[..]);
+			Ok(())
+		})
 	}
-	fn size(&self) -> u32 {
-		u32::try_from(self.len()).expect("size of Wasm linear memory is <2^32; qed")
+
+	fn size(&self) -> u64 {
+		let len = self.pages() as u64 * PAGE_SIZE as u64;
+		u64::try_from(len).expect("size of Wasm linear memory is <=2^32; qed")
 	}
 }
```

Comment:
> 2\. The `try_from` here is useless as you're converting an `u64` into an `u64` here. (:

:facepalm: :see_no_evil: 



> 1. I'd probably keep this returning `u32` (since it should never physically return anything larger than `u32` so it's a little confusing to have it return `u64`)

2^32 doesn't fit into u32. I mean we can address the entire memory, but to represent its size this is not possible. Or my brain here is doing some stupid thing?

Reply:
> 2^32 doesn't fit into u32. I mean we can address the entire memory, but to represent its size this is not possible. Or my brain here is doing some stupid thing?

Hmm... I guess this is a good point. Yeah, you're right.